### PR TITLE
[Tooltip] Fix tooltip visibility logic

### PR DESCRIPTION
**Pull request by jjenzz**
A couple of fixes here so I've left comments inline to clarify what's what.

### CODE REVIEW

File: packages/react/tooltip/src/Tooltip.tsx [hunk 1]
```diff
@@ -164,10 +163,8 @@ const Tooltip: React.FC<TooltipProps> = (props: ScopedProps<TooltipProps>) => {
         // we dispatch here so `TooltipProvider` isn't required to
         // ensure other tooltips are aware of this one opening.
         document.dispatchEvent(new CustomEvent(TOOLTIP_OPEN));
-      } else {
-        isFocusOpenRef.current = false;
+        onOpen();
       }
-      window.clearTimeout(openTimerRef.current);
```
Comment (jjenzz) [inline packages/react/tooltip/src/Tooltip.tsx:170]: This was only clearing if the `open` value changed from `false` to `true` but not if the trigger was focused and then the mouse entered it because when it was focused, open will have been true and mousing over it would have tried to set open to true again which will have ran the open timer.

File: packages/react/tooltip/src/Tooltip.tsx [hunk 2]
```diff
@@ -241,12 +243,20 @@ const TooltipTrigger = React.forwardRef<TooltipTriggerElement, TooltipTriggerPro
...
-          onBlur={composeEventHandlers(props.onBlur, context.onTriggerLeave)}
+          onMouseDown={composeEventHandlers(props.onMouseDown, () => {
+            isMouseDownRef.current = true;
+            context.onClose();
+            document.addEventListener('mouseup', () => (isMouseDownRef.current = false), {
+              once: true,
+            });
+          })}
+          onFocus={composeEventHandlers(props.onFocus, () => {
+            if (!isMouseDownRef.current) context.onOpen();
```
Comment (jjenzz) [inline packages/react/tooltip/src/Tooltip.tsx:254]: Explicitly only open if it isn't a mousedown focus.

File: packages/react/tooltip/src/Tooltip.tsx (hunk 1 above)
Comment (benoitgrelard) [reply packages/react/tooltip/src/Tooltip.tsx:170]: Good catch!

File: packages/react/tooltip/src/Tooltip.tsx [hunk 3]
```diff
@@ -164,17 +162,25 @@ const Tooltip: React.FC<TooltipProps> = (props: ScopedProps<TooltipProps>) => {
...
     },
   });
-
   const stateAttribute = React.useMemo(() => {
-    return open ? (wasOpenDelayed ? 'delayed-open' : 'instant-open') : 'closed';
-  }, [wasOpenDelayed, open]);
+    return open ? (wasOpenDelayedRef.current ? 'delayed-open' : 'instant-open') : 'closed';
+  }, [open]);
+
+  const handleOpen = React.useCallback(() => {
```
Comment (benoitgrelard) [inline packages/react/tooltip/src/Tooltip.tsx:174]: Could this be name `handleOpenInstantly` or something like that?
It's all getting a bit hard to follow now in `onTriggerEnter` with the branching + the external handler.

File: packages/react/tooltip/src/Tooltip.tsx [hunk 4]
```diff
@@ -241,12 +249,20 @@ const TooltipTrigger = React.forwardRef<TooltipTriggerElement, TooltipTriggerPro
           ref={composedTriggerRef}
           onMouseEnter={composeEventHandlers(props.onMouseEnter, context.onTriggerEnter)}
           onMouseLeave={composeEventHandlers(props.onMouseLeave, context.onTriggerLeave)}
-          onMouseDown={composeEventHandlers(props.onMouseDown, context.onTriggerLeave)}
-          onFocus={composeEventHandlers(props.onFocus, context.onTriggerFocus)}
-          onBlur={composeEventHandlers(props.onBlur, context.onTriggerLeave)}
+          onMouseDown={composeEventHandlers(props.onMouseDown, () => {
+            context.onClose();
+            isMouseDownRef.current = true;
+            document.addEventListener('mouseup', () => (isMouseDownRef.current = false), {
```
Comment (benoitgrelard) [inline packages/react/tooltip/src/Tooltip.tsx:255]: I always get confused with those, what if the trigger unmounts between mousedown and mouseup? Then it's leaking no?

File: packages/react/tooltip/src/Tooltip.tsx (hunk 3 above)
Comment (jjenzz) [reply packages/react/tooltip/src/Tooltip.tsx:174]: I feel like `handleOpen` and `handleClose` is already instant... otherwise should we rename all of our `handleOpen` handlers in our other comps?  It already implies that it's instant. 

If it helps, I've abstracted the delayed logic into a `handleDelayedOpen` so that you can see easily that sometimes it is delayed.

File: packages/react/tooltip/src/Tooltip.tsx (hunk 4 above)
Comment (jjenzz) [reply packages/react/tooltip/src/Tooltip.tsx]: Good spot. I've added a clean up for this!

File: packages/react/tooltip/src/Tooltip.tsx (hunk 3 above)
Comment (benoitgrelard) [reply packages/react/tooltip/src/Tooltip.tsx:174]: Yeah that works for me.

**Review by benoitgrelard — APPROVED** [review]
